share/wake/lib/core/lens.wake

```diff
@@ -0,0 +1,156 @@
+# Copyright 2021 SiFive, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You should have received a copy of LICENSE.Apache2 along with
+# this software. If not, you may obtain a copy at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+package wake
+
+export def (l: List a) |> (t: a => List b): List b = mapFlat t l
+
+# TODO: How do we feel about `x | f |> g` vs `|> x |> f |> g`?
+export def |> (x: a): List a = x, Nil
+
+# NOTE: These are not part of an offline discussion but I think they're a
+#       useful concept. They're a kind of horizontial composition rather
+#       than vertical. You can use them to do things like set/get a value
+#       in a fallback like way. We can drop these. getJString made me
+#       think we might want something like this for some unknown reason
+#
+# Example:
+#     def getJString j =
+#         |> j
+#         |> jString ++> (_ | jArray |> jString)
+#
+#     # The edit version is a lot more tricky to use
+#     def editJString f =
+#         jeString ++< (jeArray $ jeString $ _) $ f
+export def (t1: a => List b) ++> (t2: a => List b): a => List b =
```

Comment:
So, I DO see the need to be able to combine two edits. We know we can nest edit functions using `$`:
```
def inc3 = editPlanUsage $ editUsageMemory $ (_+3)
```
Like before, I think if you give both editors a name, you can then already compose them:
```
def bigEdit =
  editPairFirst
  $ (_ | inc3 | setPlanLabel "fun")
```